policy_agent/policy_agent.py

```diff
@@ -0,0 +1,655 @@
+                    self.pods[pod_key] = po
+                    del self.changed_pods[pod_key]
+
+        # As endpoint lists change, create/add new profiles to pods as
+        # necessary
+        # TODO: if some object not ready for resync, finish loop w/out moving to
+        # processed list
+        for ep_key, ep in self.changed_endpoints.items():
+            # Get policy of the Service and namespace. Uses same namespace/name
+            # dict key.
+            try:
+                svc_type = self.services[ep_key].type
+            except KeyError:
+                _log.warning("Service %s not yet in store" % ep_key)
+                continue
+
+            try:
+                ns_policy = self.namespaces[ep.namespace].policy
+            except KeyError:
+                _log.warning("Namespace %s not yet in store" % ep.namespace)
+                continue
+
+            _log.info("Using svc policy %s and ns policy %s" % \
+                      (svc_type, ns_policy))
+
+            if ns_policy == "Open" or svc_type == SVC_TYPE_NAMESPACE_IP:
+                # If namespace is open, or svc is closed, do nothing.
+                # Namespace_IP services can only exists in Closed namespaces,
+                # in which case NamespaceIP service policy is redundant
+                # When the Namespace is open, all valid service types are open as well.
+                self.endpoints[ep_key] = ep
+                del self.changed_endpoints[ep_key]
+            else:
+                # else define service policy
+                associations = ep.generate_svc_profiles_pods()
+                for profile, pods in associations.items():
+                    # find pods and append profiles
+                    for pod in pods:
+                        existing_pod = self.match_pod(
+                            namespace=ep.namespace, name=pod)
+                        _log.debug("Adding profile %s to pod %s" % (profile, existing_pod))
+                        if existing_pod:
+                            try:
+                                _datastore_client.append_profiles_to_endpoint(profile_names=[profile],
+                                                                              endpoint_id=existing_pod.ep_id)
+                            except ProfileAlreadyInEndpoint:
+                                _log.warning("Applying %s to Pod %s : Profile Already exists" % \
+                                             (profile, existing_pod.key))
+                        else:
+                            _log.warning("Pod %s is not yet processed" % (pod))
+
+                # declare Endpoints obj processed
+                self.epResourceVersion = ep.resourceVersion
+                self.endpoints[ep_key] = ep
+                del self.changed_endpoints[ep_key]
+
+
+class Resource():
+    """
+    Resource objects pull pertinent info from json blobs and maintain universal functions
```

Comment:
What does "maintain universal functions" mean here?  Does it mean they have methods?  That is sort of implied ;)

A more useful docstring for the Resource class might be:

```
"""
The Resource class is an abstract super class which represents Kubernetes API objects.  The class defines a number of abstract methods which must be implemented in subclasses.
"""
```